Upgrade gating

william-hutchison opened the pull request:
- Update the `gate_interactive` function to allow colour, shape, size and alpha to be defined by either a vector of values or a fixed value.
- Simplify interactive gating backend code.  
- Improve interactive gating visual interface. 
- Deprecate superseded gate functions. 
- Update documentation. 

Let me know if there are any issues or if you would like to see any changes. I am satisfied with this as the final iteration of the tidygate upgrade. Please push to CRAN when possible so that I can remove the github remote from tidySpatialExperiment.

stemangiola commented on R/methods.R:
```diff
@@ -141,6 +145,7 @@ gate_int <- function(.dim1,
                      .group_by = NULL,
                      gate_list = NULL,
                      ...) {
+  lifecycle::deprecate_warn("1.0.0", "tidygate::gate_int()")
```
Please to all deprecation, add the alternative to the message, the one function the users should use. Similarly to `sccomp`, and `tidybulk`

william-hutchison replied:
Thank you for the suggestion, I have made this change.

stemangiola commented on R/methods.R:
```diff
@@ -182,122 +187,191 @@ gate_int.numeric = 	function(  .dim1,
 #'}
 #' @export
-gate_simple <-
+gate_interactive <-
```
In the original design, we used the gate function that became programmatic if the predefined gates were used as arguments.

I think we should stick to one `gate` function, and call either the programmatic or interactive in the backend.

william-hutchison replied:
I have also been thinking about whether to combine or split these functions. I believe that by keeping these function names separate tidygate easier to use. 

When gating programmatically, the user will be able to look at ?gate_programmatic and see only arguments relative to what they are trying to do (e.g. shape and size will not be shown). The same is true for looking at the ?gate_interactive. Also, when reading code that uses tidygate, keeping two functions makes it very clear what was done. If the functions are combined, the reader will have to look for the "brush_data" argument, and know that it indicates gating was done programmatically. 

I agree it would be ideal to retain as much from the previous interface as possible, but given we are already changing a lot I think this could be an opportunity to make things clearer for the user. 

But, there is no technical reason the functions can't be combined and I would be happy to do so if you feel strongly about this.

stemangiola replied:
I understand the documentation argument, but I in my opinion the complexity of adding two switches to make one action

- changing function name if you want a different behaviour and 
- using a further argument to have a different behaviour

makes the interface much more complex

Feel free to make a little survey (I did it for tidybulk, about the interface), but in my mind I see this pretty clearly.


> the reader will have to look for the "brush_data" argument, and know that it indicates gating was done programmatically

`brush_data`, could be a self explanatory name, we could call it `programmatic_gate`

stemangiola commented on tests/testthat/test-methods.R:
```diff
@@ -82,3 +82,16 @@ test_that("gate grouping", {
+  
+  expect_equal(
+    mtcars |>
+      mutate(gate_programatic = gate_programmatic(mpg, wt, brush_data = demo_brush_data)) |>
```
Any sources where the grammar "brush" was taken from? how about the old argument name from the previous gating engine?

william-hutchison replied:
The old argument name was "gate_list", which I think no longer made sense as gate data / brush data is now stored in a single tibble. 

I chose to rename the argument to "brush_data" for clarity, given we also often call the output of the gate function / selected points "gated".  But, let me know if you have any other suggestions.

stemangiola reviewed (changes requested):
Well done. Just few points, about increasing consistency with the old front end.

william-hutchison commented:
In addition to updating the depreciation message, I also switched the gate output to a vector of strings.

william-hutchison commented:
- `gate_programmatic` and `gate_interactive` are no longer exported. They have been replaced with a single function - `gate`.
- For simplicity, additional information about selected points is no longer saved to `tidygate_env`.

stemangiola commented on README.Rmd:
```diff
@@ -6,116 +6,108 @@ output: github_document
+mtcars_gated <-
+  mtcars_gated |>
+  mutate(gated_programmatically = gate(
+    x = mpg, y = wt, programmatic_gates = tidygate_env$gates
```
Here, I would use the example of the gates environment variable being first saved as RDS and then read into the programmatic gates argument. This is because if we want reproducibility, it has to be clear that the variable should be on disk and not in memory.

stemangiola commented on README.Rmd:
```diff
@@ -6,116 +6,108 @@ output: github_document
-  mutate( gate = gate_chr( Dim1, Dim2 ) )
+mtcars_gated <- 
+  mtcars |>
+  mutate(gated_interactively = gate(x = mpg, y = wt, colour = disp))
```
If you change the column name based on interactive / programmatic, your code

```r
... |>
mutate(gated_interactively = ) |>
filter(gate_interactively) |>
... 150 lines of code

mutate(a = if_else(gate_interactively ...))
```

would fail if you change to

```r
mutate(gated_programmatic = ) |>
```

While the idea is to change one argument and reuse your code as usual.

In the README, vignettes and examples , I would not really distinguish between gate_interactively and gate_programmatically, we are adding a moving part that created complexity.

stemangiola reviewed (changes requested):
Amazing.

1) 
On this command

```r
mtcars_gated <- 
    mtcars |>
    mutate(gated_interactively = gate(x = mpg, y = wt, colour = disp, size =hp, alpha = hp))
```
I get

<img width="900" alt="image" src="https://github.com/stemangiola/tidygate/assets/7232890/669f7784-2f2c-4f25-99ae-9674b759aae3">

2) 
Also On this command


```r
mtcars_gated <- 
     mtcars |>
    mutate(gated_interactively = gate(x = mpg, y = wt, colour = disp))

```

I get this deprecation

```r
There was 1 warning in `mutate()`.
ℹ In argument: `gated_interactively = gate(x = mpg, y = wt, colour = disp)`.
Caused by warning:
! The `<scale>` argument of `guides()` cannot be `FALSE`. Use "none" instead as of ggplot2 3.3.4.
ℹ The deprecated feature was likely used in the tidygate package.
  Please report the issue at <https://github.com/stemangiola/tidygate/issues>.
This warning is displayed once every 8 hours.
Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated. 
```

But if I am using the new framework I should not get deprecation warnings.

3) after the interactive gate, when it's completed, we should print a message saying "please save <this variable...> and use it <as such...> for reproduceability" 

4) I remember, we had a discussion what variable type the gate column should be, at this moment is one or empty as a character. Have empty characters could be quite counterintuitive for filtering and regular expression, I feel it would be better to either character numbers, including "0" and "1", or probably more elegant and extensible is to return a bullion, either true or false, whether the point is within the gate or not. This would allow for defining multiple gates and using bullion operation in diverse to selected desert cells, even if overlapping two or more gates.

william-hutchison commented:
Thank you for the suggestions and spotting these issues.

1 - This is happening because size and alpha coerce the input column to a factor and hp has too many levels for the plot to handle. I have now limited factors for size, shape and alpha to 6 in the documentation and provide an informative error message when outside of this range. 

2 - This depreciation warning is caused by my use of an outdated syntax in ggplot2, I have updated the syntax. 

3 - I have added a simple message reminding the user that interactively drawn gates are temporarily saved: "tidygate says: interactively drawn gates are temporarily saved to tidygate_env$gates"

4 - I agree that returning a boolean would be more elegant, but it would not be possible to do this separately for each gate within mutate. We could either loose gate information or return some kind of tibble or list of values, but I feel both of these solutions do more harm than good. The currently returned string allows the user to very easy see which points are in which gates, and select points which appear in gates of interest, as demonstrated in the README. 

The output is currently "1" if the point is only in gate 1, but expands endlessly to "1,2,4,8" ect capturing each gate the point is in. Following this logic, I think it is best to have a point which appears in no gates as an empty string, as 0 suggests it appears within a gate called 0.

5 - I agree it would be better if the plot launches within the RStudio quadrant interface by default, if running in RStudio. This was strangely the default behaviour for me already so I cannot test, but I have tried to fix this for others. Is the default interface now the RStudio quadrant for you too? 

6 - I have added an example of saving and loading gates to the README / vignette.

7 - I have replaced the "gated_programmatic" and "gated_interactive" variable names with "gated" for storing gate function output in the documentation and README / vignette

stemangiola reviewed (changes requested):
Great.

for (1), I believe I already solved this issue in the old framework, which can accept factor or numeric. Please have a lot, there might not me need to implement limits etc.

Have a look to what I do for all those variables.

for (4) I did not realise we can draw multiple gates at once, awesome!

Number is good, the only detail is that if we leave "" then the tibble will be ackward

<img width="648" alt="image" src="https://github.com/stemangiola/tidygate/assets/7232890/fcda34c4-0e13-4872-a4c4-3732db378a1e">

If we use 0 for outside the gate the tibble is what we would expect

<img width="826" alt="image" src="https://github.com/stemangiola/tidygate/assets/7232890/4c426898-343d-4465-aaa7-ddc9a7a052f0">

Boyond this it is good to go. Very well done.

william-hutchison commented:
I have enabled numeric vector input to size and alpha parameters and updated the documentation to match.

Regarding 0s or empty strings, I agree the empty strings do look a bit odd in the tibble, but to me, the 0s are quite confusing. Another option could be to set points appearing in no gates to NA, but this also looks odd in the tibble. To me the most elegant solution is to leave the strings empty.

stemangiola commented:
> I have enabled numeric vector input to size and alpha parameters and updated the documentation to match.
> 
> Regarding 0s or empty strings, I agree the empty strings do look a bit odd in the tibble, but to me, the 0s are quite confusing. Another option could be to set points appearing in no gates to NA, but this also looks odd in the tibble. To me, the most elegant solution is to leave the strings empty.

I like NA as a solution; it does not look odd and is actually correctly non-applicable, and we don't have any gate for those points.

<img width="809" alt="image" src="https://github.com/stemangiola/tidygate/assets/7232890/49a8fe7c-2186-4767-aef1-41354f8cff93">

william-hutchison commented:
Sounds good, I have replaced empty strings with NAs.

stemangiola commented:
Thanks,

the command

```r
mtcars_gated <- 
    mtcars |>
    mutate(gated_interactively = gate(x = mpg, y = wt, colour = wt, size =hp))
```

fails, while it works for the old framework

```r
mtcars_gated <- 
    mtcars |>
    mutate(gated_interactively = gate_chr(.dim1 = mpg,  .dim2 = wt, .color = wt, .size = hp))
```

<img width="523" alt="image" src="https://github.com/stemangiola/tidygate/assets/7232890/00811a91-3af5-4de4-9261-cdda98482d2a">

Please adapt the old-framework strategy to the argument as much as you can, I think we risk to be reinventing the weal.

Also a comment on the thresholds of the argument (when continuous). `ggplot` does not impose thresholds for the user, eliminating complexity, and the need to read manual.

```r
tibble(a = 1:10, b = 1:10) |> ggplot(aes(a, b, alpha = a, size = b)) + geom_point()
```

<img width="573" alt="image" src="https://github.com/stemangiola/tidygate/assets/7232890/524a248f-b95d-4427-b1fb-1cf00f65c4c0">

We should not either. Remember do not expose backend, technical needs to the user. The API should do the heavy lifting not the user.

william-hutchison commented:
I mistakenly thought ggplot used the actual values supplied to size and alpha which is why I imposed the limits. I realise now this is not the case and have removed the limits, which caused the error. Thanks for the suggestion. Unfortunately plotly seems to be unable to display the continuous size and alpha legends, so we might have to proceed without these for the moment.

stemangiola commented:
> I mistakenly thought ggplot used the actual values supplied to size and alpha which is why I imposed the limits. I realise now this is not the case and have removed the limits, which caused the error. Thanks for the suggestion. Unfortunately plotly seems to be unable to display the continuous size and alpha legends, so we might have to proceed without these for the moment.

Sure, legends are not the end of the world.
